lib/registry/src/graphql.rs

```diff
@@ -0,0 +1,251 @@
+    form_modifier: F,
+) -> anyhow::Result<()>
+where
+    V: serde::Serialize,
+    F: FnOnce(Form) -> Form,
+{
+    let client = {
+        let builder = Client::builder();
+
+        #[cfg(not(target_os = "wasi"))]
+        let builder = if let Some(proxy) = proxy::maybe_set_up_proxy()? {
+            builder.proxy(proxy)
+        } else {
+            builder
+        };
+        builder.build()?
+    };
+
+    let vars = serde_json::to_string(&query.variables).unwrap();
+
+    let form = Form::new()
+        .text("query", query.query.to_string())
+        .text("operationName", query.operation_name.to_string())
+        .text("variables", vars);
+
+    let form = form_modifier(form);
+
+    let user_agent = format!(
+        "wapm/{} {} {}",
+        env!("CARGO_PKG_VERSION"),
+        whoami::platform(),
+        whoami_distro(),
+    );
+
+    let mut res = client
+        .post(registry_url)
+        .multipart(form)
+        .bearer_auth(env::var("WAPM_REGISTRY_TOKEN").unwrap_or_else(|_| login_token.to_string()))
+        .header(USER_AGENT, user_agent);
+
+    if let Some(t) = timeout {
+        res = res.timeout(t);
+    }
+
+    let res = res.send()?;
+
+    let _: Response<serde_json::Value> = res.json()?;
+
+    Ok(())
+}
+
+pub fn execute_query_modifier_inner<R, V, F>(
+    registry_url: &str,
+    login_token: &str,
+    query: &QueryBody<V>,
+    timeout: Option<Duration>,
+    form_modifier: F,
+) -> anyhow::Result<R>
+where
+    for<'de> R: serde::Deserialize<'de>,
+    V: serde::Serialize,
+    F: FnOnce(Form) -> Form,
+{
+    let client = {
+        let builder = Client::builder();
+
+        #[cfg(not(target_os = "wasi"))]
+        let builder = if let Some(proxy) = proxy::maybe_set_up_proxy()? {
+            builder.proxy(proxy)
+        } else {
+            builder
+        };
+        builder.build()?
+    };
+
+    let vars = serde_json::to_string(&query.variables).unwrap();
+
+    let form = Form::new()
+        .text("query", query.query.to_string())
+        .text("operationName", query.operation_name.to_string())
+        .text("variables", vars);
+
+    let form = form_modifier(form);
+
+    let user_agent = format!(
+        "wapm/{} {} {}",
+        env!("CARGO_PKG_VERSION"),
+        whoami::platform(),
+        whoami_distro(),
+    );
+
+    let mut res = client
+        .post(registry_url)
+        .multipart(form)
+        .bearer_auth(env::var("WAPM_REGISTRY_TOKEN").unwrap_or_else(|_| login_token.to_string()))
+        .header(USER_AGENT, user_agent);
+
+    if let Some(t) = timeout {
+        res = res.timeout(t);
+    }
+
+    let res = res.send()?;
+    let response_body: Response<R> = res.json()?;
+    if let Some(errors) = response_body.errors {
+        let error_messages: Vec<String> = errors.into_iter().map(|err| err.message).collect();
+        return Err(anyhow::anyhow!("{}", error_messages.join(", ")));
+    }
+    Ok(response_body.data.expect("missing response data"))
```

Comment:
Under what circumstances would we typically get a missing response data? If this is a generic utility used when executing any GraphQL query, it might be useful to turn this into an error.

Reply:
No, I don't think it makes sense to strongly type this error: it only fails if - for example - the backend is down and isn't returning any data.

Reply:
Sorry, I just saw that you just wanted me to turn this into an error: yeah, that's okay